design/ceph/object/store.md

```diff
@@ -93,6 +93,7 @@ The gateway settings correspond to the RGW service.
 - `securePort`: The service port where the RGW service will be listening (https)
 - `instances`: The number of RGW pods that will be started for this object store
 - `placement`: The rgw pods can be given standard Kubernetes placement restrictions with `nodeAffinity`, `tolerations`, `podAffinity`, `podAntiAffinity`, and `topologySpreadConstraints` similar to placement defined for daemons configured by the [cluster CRD](/deploy/examples/cluster.yaml).
+- `hostnames`: A list of hostnames to use for accessing the bucket directly like a subdomain in the endpoint. For example if the default service endpoint `rook-ceph-rgw-ocs-storagecluster-cephobjectstore.openshift-storage.svc:80` added and objectstore contains a bucket name `sample`, then s3 style dns domain would be `http://sample.rook-ceph-rgw-ocs-storagecluster-cephobjectstore.openshift-storage.svc:80`
```

Comment:
Let's use upstream example names
```suggestion
- `hostnames`: A list of hostnames to use for accessing the bucket directly like a subdomain in the endpoint. For example if the default service endpoint `rook-ceph-rgw-my-store.rook-ceph.svc:80` added and the object store contains a bucket named `sample`, then the s3 style dns domain would be `http://sample.rook-ceph-rgw-my-store.rook-ceph.svc:80`
```

Reply:
To have the DNS subdomain style for buckets, every bucket will require an entry in this list of `hostnames`, correct? Could we automate this with another setting? For example, if there was a setting `enableDNSSubdomains`, then every time a COSI bucket is provisioned, we could automatically create the subdomain for that bucket?

Reply:
I like Travis' idea here. If this feature is for per-bucket access, then it seems to me like Rook should either create an endpoint for every bucket created via OBC, or (and?) Rook should allow users to specify the bucket that is addressed via the endpoint. Or is the bucket that is referenced always going to be `<bucket-name>.<subdomain>`?

Reply:
Also, I have to say that `hostnames` seems like a confusing choice of naming convention. I think for advanced features, it is probably easiest for users if the terminology Rook uses matches 1-to-1 with terminology that Ceph uses.

I'm looking at the red hat doc here: https://access.redhat.com/documentation/en-us/red_hat_ceph_storage/2/html/object_gateway_guide_for_ubuntu/configuration#configuring_the_dns)

```
[client.rgw.<STATIC-SITE-HOSTNAME>]
...
rgw_enable_static_website = true
rgw_enable_apis = s3website
rgw_dns_name = objects-zonegroup.domain.com
rgw_dns_s3website_name = objects-website-zonegroup.domain.com
rgw_resolve_cname = true
...
```

I don't see "hostname" only as "static site hostname", but there are many configurables, and this design doc does not adequately detail what Rook should actually be doing.

From looking at this doc, maybe a struct like this is needed per-RGW.
```
staticWebHosting:
  enabled: <bool>
  siteHostname: <string>
  dnsName: <string> 
  dnsS3WebsiteName: <string>
```

But what about multisite? What about the multiple-objectstore-per-zone setups? Should these configs be on the zone? Or, since these are applied to the zonegroup, should they be on the zonegroup itself?

I also don't understand how `rgw_dns_name` fits into the equation. Is this actually a subdomain that should go in the config? Is it a per-RGW hostname? Is it a load balancer? I think this is a failing of the Ceph docs that it is unclear how to use the feature to accomplish the end goal.

Ultimately, I think Travis and I need quite a bit more information to help fit this Ceph feature into Rook's architecture. I think you have ideas about what needs to be implemented to achieve this feature, but they aren't all communicated in the design. This is a shortcoming that I have observed with designs in the past. 

Regardless of who is planning a new, complex feature, Travis and I need the design to break the design into parts and detail them clearly so that we can consider all aspects of the planning for Rook. We've seen that even simple features can create long-lasting code maintenance challenges if they are not planned and implemented carefully, and the work that needs to be done for RGW seems to only be complicated these days. If a feature isn't planned carefully, code reviews take a very long time, everyone gets exhausted, and all sorts of issues are more likely to make it into code. We really want to avoid all of that, and that is why we are so insistent on having better quality designs before work begins.

Please consider this here, and for all future designs as well. Design work is quite challenging, but it is critical for long-term success of projects and features. 

With every design, we need to understand:
1. what the end goal of the feature is, as understood by users
2. how Ceph needs to be configured to achieve the user's end goals
3. what the scope of the Ceph changes is
    * per-bucket, per-RGW, per-objectstore, per-zone, per-zonegroup, etc.
    * there may be multiple scopes
5. a detailed plan for CRD configuration changes
    * this is the user-facing API, and it's critical that we plan for forward and backward compatibility, otherwise future changes can be awkward
6. a delineation of what Rook will do automatically and what users will do themselves
7. a plan for breaking up the work into sub-parts (if possible)
    * smaller changesets will make it through code review faster and with less exhaustion
    * e.g., users will create this thing in the first implementation, then Rook will do it automatically in a follow-up
8. noteworthy alternative designs that were rejected, and why they were rejected
9. drawbacks and limitations of the current design
   * e.g., corner cases that are unsupported, or future design work needed

For object store, I think we are also always concerned about how new features affect multi-site. Especially the use-case that I mentioned where a user might run 2 CephObjectStores for 1 CephObjectZone where one store is used for client traffic and the other store is used for multisite replication traffic. This stands out to me as one of the most complex scenarios, so I think it's a good one to examine for every design.
A user helped enable this usage pattern here: https://github.com/rook/rook/pull/12560

Reply:
@thotz please address this comment in full. I see that you marked it as resolved without addressing the items, and I spent a lot of time to lay out clear expectations and explain why they are important.

Reply:
> * his is the user-facing API, and it's critical that we plan for forward and backward compatibility, otherwise future changes can be awkward

Please note the doc which you are referring is RHCS 2.0 earlier version of ceph, in the latest docs RHCS 6.0, these steps are removed and the hostname configuration in the zonegroup and configure DNS server separately to the dynamic bucket endpoint using wildcard entry. Nothing additional configured in the RGW side. Naming part even I am not convinced much. But its is transparent with zonegroup configuration n Ceph

Reply:
`hostname` is really not a unique way of identifying this feature, even if it is transparent with zonegroup. The word "hostname" conflicts with so many concepts in so many parts of Ceph, Rook, and Kubernetes that I really don't think we should use it unless it's as a sub-field of some other struct. Also, Rook configures an "object store" that has RGWs, and the zone group JSON is abstracted away for the most part.

I'm still really struggling with what is happening in this design, and I'm very confused. The RHCS 6 docs don't even say what the feature is called. Only "Add a wildcard to the DNS" with the description "You can add the wildcard such as hostname to the DNS record of the DNS server." That doesn't really explain to me what the feature is supposed to do, only how to achieve something that has no clear purpose to me.

It also tells me nothing about buckets, which is part of the title of this PR.

However, the "Static Web Hosting" docs in another section do talk about buckets, and that doc matches what I wrote 3 weeks ago. https://access.redhat.com/documentation/en-us/red_hat_ceph_storage/6/html/object_gateway_guide/basic-configuration#static-web-hosting-gateway-setup-rgw

The [original issue](https://github.com/rook/rook/issues/4780) that this design is purporting to solve specifically mentions the `rgw-dns-name` field which is documented at the "Static Web Hosting" link I just gave.

I am, at this point, extremely confused and generally convinced that this design is not addressing what was requested in https://github.com/rook/rook/issues/4780.